src/back/spv/writer.rs

```diff
@@ -901,49 +901,111 @@ impl Writer {
                         let left_ty_inner = &ir_module.types[left_ty.unwrap()].inner;
                         let right_ty_inner = &ir_module.types[right_ty.unwrap()].inner;
 
-                        let (result_type_id, vector_id, scalar_id) =
-                            match (left_ty_inner, right_ty_inner) {
-                                (
-                                    crate::TypeInner::Vector { .. },
-                                    crate::TypeInner::Scalar { .. },
-                                ) => (
-                                    self.get_type_id(
-                                        &ir_module.types,
-                                        LookupType::Handle(left_ty.unwrap()),
-                                    ),
-                                    left_id,
-                                    right_id,
-                                ),
-                                (
-                                    crate::TypeInner::Scalar { .. },
-                                    crate::TypeInner::Vector { .. },
-                                ) => (
-                                    self.get_type_id(
-                                        &ir_module.types,
-                                        LookupType::Handle(right_ty.unwrap()),
-                                    ),
-                                    right_id,
-                                    left_id,
-                                ),
-                                _ => unreachable!("Expression requires both a scalar and vector"),
-                            };
+                        let left_result_type_id = self
+                            .get_type_id(&ir_module.types, LookupType::Handle(left_ty.unwrap()));
 
-                        let load_id = self.generate_id();
-                        let load_instruction = super::instructions::instruction_load(
-                            result_type_id,
-                            load_id,
-                            vector_id,
+                        let right_result_type_id = self
+                            .get_type_id(&ir_module.types, LookupType::Handle(right_ty.unwrap()));
+
+                        // Load left- and right-hand side via OpLoad
+                        let left_load_id = self.generate_id();
+                        let right_load_id = self.generate_id();
+
+                        block.body.push(super::instructions::instruction_load(
```

Comment:
why are there `load` instructions here?

Reply:
That is, because the values are from an `OpVariable` for instance, and to use an `OpVariable`, you have to load them through `OpLoad`.

Reply:
hmm, but there isn't any check here about whether the target is OpVariable or not. So how do we know that unconditionally "load"-ing is the right thing to do?

Reply:
That is true, but as far as I know there are only two ways to use binary instructions at high level: via pointers, and via constants.

I have done some research at http://shader-playground.timjones.io/, and when doing it via pointers, there is always an `OpLoad` needed.

GLSL:
```glsl
void main()
{
    vec2 vec_1 = vec2(1.0, 1.0);
    float float_1 = vec_1.x * vec_1.y;
    float float_3 = float_1 * float_1;
}
```
This will always translate back to `OpLoad`, whether it is via an `OpAccessChain` or `OpVariable`. There probably are more ways of accessing value via pointers, but all pointers always need to be loaded in via OpLoad to be useful (as far as I know).

What is good to see, is that when I am using constants, instead of variables at the shader playground, it pre-calculates the value, and uses it directly (which is allowed).

GLSL:
```glsl
void main()
{
    float float_1 = 1.0 * 2.0;
}
```

However, because of the pre-calculation, the resulting SPIR-V output does not even contain a binary instruction, as it is already calculated.

I think it is safe to say for now to always use `OpLoad` here, but maybe you think about a case where this might not be, please let me know, so we can reconsider this piece of code.

Reply:
We can't *always* use `OpLoad` there. It should only happen if the target expression is a variable. Once it's a different expression, there should be no load. For example, if you do `float x = 2 * (y + 1);`

Reply:
Ah yes, I understand what you mean. I have a rough idea, and will try to implement it tomorrow: what if we never check for the specific `Expression::LocalVariable`, but let the match arm of `Expression::LocalVariable` always create and return an `OpLoad`? 

That should fix this issue, and the rest of the code will be cleaner, as [`Expression::Return`](https://github.com/gfx-rs/naga/blob/master/src/back/spv/writer.rs#L1147), and [`Expression::Store`](https://github.com/gfx-rs/naga/blob/master/src/back/spv/writer.rs#L1191) have a specific match for if the id of the expression is of the expression type `Expression::LocalVariable`. This is pretty dirty to be honest, but I didn't see that back then.

Reply:
yes, both `LocalVariable` and `GlobalVariable` variants need to generate `OpLoad`.

Reply:
Fixed and now it supports even "unlimited" expression nesting:
```
var float_1: f32 = 2.0 * (2.0 * (4.0 * 2.0));
```

Giving the predicted result without an `OpLoad`, like you said (and correctly validates):
```
%5 = OpVariable %_ptr_Function_float Function
%13 = OpFMul %float %float_4 %float_2
%12 = OpFMul %float %float_2 %13
%11 = OpFMul %float %float_2 %12
      OpStore %5 %11 None
```